File: Autonomous Mobile Robots/Chapter1/Lesson3/Chapter1_Lesson3.java
/*
Chapter 1 — Lesson 3: Environment Representations for AMR
File: Chapter1_Lesson3.java

Java implementation of:
  1) Occupancy grid (rectangles)
  2) Brushfire distance transform (Manhattan, multi-source BFS)
  3) Signed distance field (SDF) and disc-robot inflation

Compile:
  javac Chapter1_Lesson3.java
Run:
  java Chapter1_Lesson3
*/

import java.io.FileWriter;
import java.io.IOException;

public class Chapter1_Lesson3 {

    static int idx(int r, int c, int W) { return r * W + c; }

    static void fillRect(byte[] occ, int H, int W, int r0, int c0, int r1, int c1) {
        if (r0 < 0) r0 = 0;
        if (c0 < 0) c0 = 0;
        if (r1 >= H) r1 = H - 1;
        if (c1 >= W) c1 = W - 1;
        for (int r = r0; r <= r1; r++) {
            for (int c = c0; c <= c1; c++) {
                occ[idx(r, c, W)] = 1;
            }
        }
    }

    static void brushfireManhattan(byte[] occ, int H, int W, int[] dist) {
        final int INF = 1000000000;
        final int N = H * W;
        for (int i = 0; i < N; i++) dist[i] = INF;

        int[] q = new int[N];
        int head = 0, tail = 0;

        for (int r = 0; r < H; r++) {
            for (int c = 0; c < W; c++) {
                int k = idx(r, c, W);
                if (occ[k] == 1) {
                    dist[k] = 0;
                    q[tail++] = k;
                }
            }
        }

        int[] dr = new int[]{-1, 1, 0, 0};
        int[] dc = new int[]{0, 0, -1, 1};

        while (head < tail) {
            int k = q[head++];
            int r = k / W;
            int c = k - r * W;
            int d = dist[k];

            for (int t = 0; t < 4; t++) {
                int rr = r + dr[t];
                int cc = c + dc[t];
                if (rr >= 0 && rr < H && cc >= 0 && cc < W) {
                    int kk = idx(rr, cc, W);
                    if (dist[kk] > d + 1) {
                        dist[kk] = d + 1;
                        q[tail++] = kk;
                    }
                }
            }
        }
    }

    static void writeCsv(String filename, byte[] grid, int H, int W) throws IOException {
        FileWriter fw = new FileWriter(filename);
        for (int r = 0; r < H; r++) {
            for (int c = 0; c < W; c++) {
                fw.write(Integer.toString(grid[idx(r, c, W)]));
                if (c + 1 < W) fw.write(",");
            }
            fw.write("\n");
        }
        fw.close();
    }

    public static void main(String[] args) throws Exception {
        final int H = 160;
        final int W = 200;
        final double res = 0.05;

        byte[] occ = new byte[H * W];
        byte[] occInv = new byte[H * W];
        int[] dToObs = new int[H * W];
        int[] dToFree = new int[H * W];
        double[] sdf = new double[H * W];
        byte[] occInfl = new byte[H * W];

        // Obstacles
        fillRect(occ, H, W, 30, 40, 55, 100);
        fillRect(occ, H, W, 90, 120, 130, 180);
        fillRect(occ, H, W, 110, 20, 140, 45);

        // Distances
        brushfireManhattan(occ, H, W, dToObs);
        for (int i = 0; i < H * W; i++) occInv[i] = (byte)(1 - occ[i]);
        brushfireManhattan(occInv, H, W, dToFree);

        // SDF
        for (int i = 0; i < H * W; i++) {
            if (occ[i] == 1) sdf[i] = -((double)dToFree[i]) * res;
            else sdf[i] = ((double)dToObs[i]) * res;
        }

        // Inflation
        final double robotRadius = 0.30;
        for (int i = 0; i < H * W; i++) {
            occInfl[i] = (byte)(sdf[i] < robotRadius ? 1 : 0);
        }

        // CSV export
        writeCsv("occupancy_grid_java.csv", occ, H, W);
        writeCsv("inflated_occupancy_grid_java.csv", occInfl, H, W);

        // Summary
        int free0 = 0, free1 = 0;
        for (int i = 0; i < H * W; i++) {
            if (occ[i] == 0) free0++;
            if (occInfl[i] == 0) free1++;
        }

        System.out.println("Grid: H=" + H + ", W=" + W + ", res=" + res + " m");
        System.out.println("Free ratio (original): " + ((double)free0 / (double)(H * W)));
        System.out.println("Free ratio (inflated): " + ((double)free1 / (double)(H * W)));
        System.out.println("Wrote: occupancy_grid_java.csv, inflated_occupancy_grid_java.csv");
    }
}

File: Autonomous Mobile Robots/Chapter1/Lesson3/Chapter1_Lesson3.m
% Chapter 1 — Lesson 3: Environment Representations for AMR
% File: Chapter1_Lesson3.m
%
% This script demonstrates:
%   1) Raster occupancy grid using rectangle obstacles
%   2) Signed distance field via bwdist
%   3) Disc-robot inflation in configuration space
%   4) Programmatic creation of a small Simulink model skeleton
%
% Requirements:
%   - Image Processing Toolbox for bwdist

clear; clc;

% Grid
H = 160; W = 200;
res = 0.05; % meters per cell

occ = zeros(H, W, 'uint8');

% Rectangle obstacles [r0 c0 r1 c1] inclusive
rects = [
    30  40  55 100
    90 120 130 180
   110  20 140  45
];

for k = 1:size(rects,1)
    r0 = max(1, rects(k,1)); c0 = max(1, rects(k,2));
    r1 = min(H, rects(k,3)); c1 = min(W, rects(k,4));
    occ(r0:r1, c0:c1) = 1;
end

occ_bool = occ == 1;
free_bool = ~occ_bool;

% Distance to obstacles for free cells
d_free_to_obs = bwdist(occ_bool) * res;

% Distance to free for obstacle cells
d_obs_to_free = bwdist(free_bool) * res;

% Signed distance field (SDF): positive outside, negative inside
sdf = double(d_free_to_obs);
sdf(occ_bool) = -double(d_obs_to_free(occ_bool));

% Disc-robot inflation
robot_radius = 0.30;
occ_infl = uint8(sdf < robot_radius);

% Export CSV
writematrix(occ, 'occupancy_grid_matlab.csv');
writematrix(occ_infl, 'inflated_occupancy_grid_matlab.csv');
writematrix(sdf, 'signed_distance_field_matlab.csv');

fprintf('Grid: H=%d, W=%d, res=%.3f m\n', H, W, res);
fprintf('Free ratio (original): %.4f\n', mean(occ(:)==0));
fprintf('Free ratio (inflated): %.4f\n', mean(occ_infl(:)==0));
fprintf('Wrote: occupancy_grid_matlab.csv, inflated_occupancy_grid_matlab.csv, signed_distance_field_matlab.csv\n');

% Optional visualization (comment out if you do not want figures)
figure; imagesc(occ); axis image; title('Occupancy grid (1=occupied)'); colorbar;
figure; imagesc(sdf); axis image; title('Signed distance field (meters)'); colorbar;

% -------------------------------------------------------------------------
% Simulink skeleton: create a model with a Constant block feeding a MATLAB
% Function block that computes a distance transform. This is a minimal
% starting point, not a complete navigation pipeline.
% -------------------------------------------------------------------------
try
    mdl = 'Chapter1_Lesson3_SDF_Model';
    if bdIsLoaded(mdl)
        close_system(mdl, 0);
    end
    new_system(mdl);
    open_system(mdl);

    add_block('simulink/Sources/Constant', [mdl '/OccupancyGrid'], ...
        'Value', 'occ', 'Position', [50 60 200 110]);

    add_block('simulink/User-Defined Functions/MATLAB Function', [mdl '/ComputeSDF'], ...
        'Position', [260 50 430 130]);

    add_block('simulink/Sinks/To Workspace', [mdl '/SDF_to_WS'], ...
        'VariableName', 'sdf_sim', 'Position', [500 60 650 110]);

    add_line(mdl, 'OccupancyGrid/1', 'ComputeSDF/1');
    add_line(mdl, 'ComputeSDF/1', 'SDF_to_WS/1');

    % Set MATLAB Function block code
    fcnPath = [mdl '/ComputeSDF'];
    set_param(fcnPath, 'Script', sprintf([ ...
        'function sdf_out = fcn(occ_in)\n' ...
        '%% occ_in: uint8 matrix, 1=occupied\n' ...
        'occ_bool = occ_in == 1;\n' ...
        'free_bool = ~occ_bool;\n' ...
        'd_free_to_obs = bwdist(occ_bool);\n' ...
        'd_obs_to_free = bwdist(free_bool);\n' ...
        'sdf_out = double(d_free_to_obs);\n' ...
        'sdf_out(occ_bool) = -double(d_obs_to_free(occ_bool));\n' ...
        'end\n' ...
    ]));

    save_system(mdl);
    fprintf('Created Simulink model: %s.slx\n', mdl);
catch ME
    fprintf('Simulink model creation skipped (reason: %s)\n', ME.message);
end
